src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java

```diff
@@ -121,37 +121,39 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
 				continue;
 			}
 
-			// Prepare data set
-			List<HtmlParameter> formDataSet = prepareFormDataSet(form.getFormFields());
+			FormData formData = prepareFormDataSet(form.getFormFields());
 
 			// Process the case of a POST method
 			if (method != null && method.trim().equalsIgnoreCase(METHOD_POST)) {
-				String query = "";
+				// Build the absolute canonical URL
+				String fullURL = URLCanonicalizer.getCanonicalURL(action, baseURL);
+				if (fullURL == null) {
+					return false;
+				}
+				log.debug("Canonical URL constructed using '" + action + "': " + fullURL);
 
 				/*
 				 * Ignore encoding, as we will not POST files anyway, so using
 				 * "application/x-www-form-urlencoded" is adequate
 				 */
 				// String encoding = form.getAttributeValue("enctype");
 				// if (encoding != null && encoding.equals("multipart/form-data"))
-				query = buildEncodedUrlQuery(formDataSet);
-				log.debug("Submiting form with POST method and message body with form parameters (normal encoding): "
-						+ query);
-
-				// Build the absolute canonical URL
-				String fullURL = URLCanonicalizer.getCanonicalURL(action, baseURL);
-				if (fullURL == null) {
-					return false;
+				String baseRequestBody = buildEncodedUrlQuery(formData.getFields());
+				if (formData.getSubmitFields().isEmpty()) {
+					notifyPostResourceFound(message, depth, fullURL, baseRequestBody);
+					continue;
 				}
 
-				log.debug("Canonical URL constructed using '" + action + "': " + fullURL);
-				notifyListenersPostResourceFound(message, depth + 1, fullURL, query);
+				for (HtmlParameter submitField : formData.getSubmitFields()) {
+					notifyPostResourceFound(
+							message,
+							depth,
+							fullURL,
+							appendEncodedUrlQueryParameter(baseRequestBody, submitField));
```

Comment:
I don't mind using other formatter rules that don't do that (using that one for consistency).